### PR TITLE
Init project

### PR DESCRIPTION
**Pull request by forhadmethun**

### CODE REVIEW

File: src/main/scala/models/Messages.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import io.circe.generic.semiauto._
+import io.circe.syntax.EncoderOps
+
+object Messages {
+  final case class InitMessageResponseData(orbs: List[Orb])
+      derives Codec.AsObject
+  final case class InitMessageResponse(
+      `type`: String,
+      data: InitMessageResponseData
+  ) derives Codec.AsObject
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Messages.scala]:
```scala
enum Response derives Codec.Enum:
  case Init(data: InitMessageResponseData)
  case Sth
```

or
```scala
trait Response

final case class Init() extends Response
```

```scala
trait Request // or. 
enum Request
```

See https://github.com/circe/circe/pull/1800 for Circe Scala 3 derivation

File: src/main/scala/models/Messages.scala [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import io.circe._
+import io.circe.generic.semiauto._
+import io.circe.syntax.EncoderOps
+
+object Messages {
+  final case class InitMessageResponseData(orbs: List[Orb])
+      derives Codec.AsObject
+  final case class InitMessageResponse(
+      `type`: String,
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Messages.scala]: You can use `Configuration.default.withDiscriminator`

File: build.sbt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    name := "agar-winner",
+    version := projectVersion,
+    scalaVersion := scala3Version,
+    libraryDependencies ++= Seq(
+      "org.scalameta" %% "munit" % "0.7.29" % Test,
+      "org.http4s" %% "http4s-ember-server" % Http4sVersion,
+      "org.http4s" %% "http4s-circe" % Http4sVersion,
+      "org.http4s" %% "http4s-dsl" % Http4sVersion,
+      "io.circe" %% "circe-core" % CirceVersion,
+      "io.circe" %% "circe-generic" % CirceVersion,
```
Comment (grzegorz-bielski) [inline build.sbt:19]: You might not need it

File: src/main/scala/models/Orb.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+case class Orb(color: String, locX: Int, locY: Int, radius: Int)
+    derives Codec.AsObject
+
+object Orb {
+  def apply(worldWidth: Int, worldHeight: Int): Orb = {
+    val color = getRandomColor()
+    val locX = scala.util.Random.nextInt(worldWidth)
+    val locY = scala.util.Random.nextInt(worldHeight)
+    val radius = 5
+    Orb(color, locX, locY, radius)
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Orb.scala]: Use IO for side effect such as Random

File: src/main/scala/models/Orb.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  private def getRandomColor(): String = {
+    val r = scala.util.Random.nextInt(151) + 50
+    val g = scala.util.Random.nextInt(151) + 50
+    val b = scala.util.Random.nextInt(151) + 50
+    s"rgb($r, $g, $b)"
+  }
+
+  def generateOrbs(settings: Settings): List[Orb] = {
+    List.fill(settings.defaultOrbs)(
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Orb.scala]: Try changing to Vector or ArraySeq

File: src/main/scala/models/Player.scala [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  val color: String = getRandomColor()
+  var score: Int = 0
+  var orbsAbsorbed: Int = 0
+
+  private def getRandomColor(): String = {
+    val r = scala.util.Random.nextInt(150) + 50
+    val g = scala.util.Random.nextInt(150) + 50
+    val b = scala.util.Random.nextInt(150) + 50
+    s"rgb($r,$g,$b)"
+  }
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Player.scala:46]:
```scala
object PlayerData:
  def createPlayer(...): IO[PlayerData]
 ```

File: src/main/scala/models/Player.scala [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+case class PlayerConfig(defaultSpeed: Int, defaultZoom: Double)
+
+case class PlayerData(playerName: String, settings: Settings) {
+  val uid: String = java.util.UUID.randomUUID().toString
+  val locX: Int = scala.util.Random.nextInt(settings.worldWidth) + 100
+  val locY: Int = scala.util.Random.nextInt(settings.worldHeight) + 100
+  val radius: Int = settings.defaultSize
+  val color: String = getRandomColor()
+  var score: Int = 0
+  var orbsAbsorbed: Int = 0
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Player.scala]:
```scala
final case class PlayerData(score: Int, ....): 
  def increaseScore: PlayerData = copy(
       score = score + 1
   )
```

File: src/main/scala/models/Settings.scala [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package models
+
+case class Settings(
+    defaultOrbs: Int = 50,
+    defaultSpeed: Int = 6,
+    defaultSize: Int = 6,
+    defaultZoom: Double = 1.5,
+    worldWidth: Int = 1440,
+    worldHeight: Int = 1440
+)
```
Comment (grzegorz-bielski) [inline src/main/scala/models/Settings.scala]: Later try `pureconfig` or `Ciris`

File: src/main/scala/routes/GameRoutes.scala [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+final class GameRoutes[F[_]](gameService: GameService[F]) extends Http4sDsl[F]:
+  def wsRoutes(using Async[F])(builder: WebSocketBuilder2[F]): HttpRoutes[F] =
+    HttpRoutes.of[F] { case GET -> Root =>
+      val out: Stream[F, WebSocketFrame] = gameService.subscribe.map(msg => {
+        WebSocketFrame.Text(msg.content)
+      })
+
+      val in: Pipe[F, WebSocketFrame, Unit] =
+        _.evalMap(a => Sync[F].delay(scribe.info(a.toString)).as(a))
```
Comment (grzegorz-bielski) [inline src/main/scala/routes/GameRoutes.scala]: You can check https://typelevel.org/log4cats/ or https://github.com/evolution-gaming/cats-helper

Comment (grzegorz-bielski) [inline src/main/scala/routes/GameRoutes.scala]: `evalTap`

File: src/main/scala/services/GameService.scala [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  def publish(m: GameMessage)(using MonadThrow[F]): F[Unit] =
+    game.flatMap(_.publish(m))
+
+  protected def deamon(using Concurrent[F]): F[Unit] =
+    def next: F[Unit] = {
+      game.flatMap(game => {
+        Spawn[F].start(game.deamon).void
+      })
+    }
+    Spawn[F].start(next).void
```
Comment (grzegorz-bielski) [inline src/main/scala/services/GameService.scala]: You could also try
```
Spawn[F].background
```

File: src/main/scala/services/GameService.scala [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      .drain
+
+object Game:
+  def create[F[_]: Concurrent]: F[Game[F]] =
+    for
+      q <- Queue.unbounded[F, Option[GameMessage]]
+      topic <- Topic[F, GameMessage]
+    yield Game(topic, q)
+
+final case class GameMessage(id: String, content: String) derives Codec.AsObject
```
Comment (grzegorz-bielski) [inline src/main/scala/services/GameService.scala]: You don't codec here, probably

File: src/main/scala/models/Messages.scala (hunk 1 above)
Comment (forhadmethun) [reply src/main/scala/models/Messages.scala]: Segregated Request and Response messages with sealed trait 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/models/Messages.scala (hunk 2 above)
Comment (forhadmethun) [reply src/main/scala/models/Messages.scala]: Renamed the variable with `messageType` to keep it simpler
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: build.sbt (hunk 3 above)
Comment (forhadmethun) [reply build.sbt:19]: It's actually necessary to get a generic decoder for the following portion https://github.com/forhadmethun/agar-winner/blob/f6ebfd8edf9498933921e6c56dae6e010a08d839/src/main/scala/services/GameService.scala#L41

File: src/main/scala/models/Orb.scala (hunk 4 above)
Comment (forhadmethun) [reply src/main/scala/models/Orb.scala]: Updated the Orb.scala with effect :) 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/models/Orb.scala (hunk 5 above)
Comment (forhadmethun) [reply src/main/scala/models/Orb.scala]: Updated with Vector 
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/models/Player.scala (hunk 6 above)
Comment (forhadmethun) [reply src/main/scala/models/Player.scala:46]: Updated the Player.scala with effect
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/models/Player.scala (hunk 7 above)
Comment (forhadmethun) [reply src/main/scala/models/Player.scala]: Added method for such case
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/models/Settings.scala (hunk 8 above)
Comment (forhadmethun) [reply src/main/scala/models/Settings.scala]: Will do in later phase :)

File: src/main/scala/routes/GameRoutes.scala (hunk 9 above)
Comment (forhadmethun) [reply src/main/scala/routes/GameRoutes.scala]: Added log4cats
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

Comment (forhadmethun) [reply src/main/scala/routes/GameRoutes.scala]: Updated with evalTap
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/services/GameService.scala (hunk 10 above)
Comment (forhadmethun) [reply src/main/scala/services/GameService.scala]: For now, I'm not changing it, so moving forward with the start method but with single fiber
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

File: src/main/scala/services/GameService.scala (hunk 11 above)
Comment (forhadmethun) [reply src/main/scala/services/GameService.scala]: Removed this unnecessary codec
ref: [changes here](https://github.com/forhadmethun/agar-winner/pull/2/files)

**Pull request merged**